File: src/shared/roadtrip/hook-scene.ts
/**
 * Handing a Road Trip hook to the Studio, as an intro SCENE.
 *
 * The workflow this closes: a clip is graded and given its telemetry overlay
 * in the Studio, a day badge is composed in Road Trip, and the two used to
 * leave as separate files to be joined on a phone. They do not have to be.
 * The badge is already an `OverlayElement[]`, the Studio already knows how to
 * hold a group of elements for the first seconds of a clip (a scene), and its
 * export already burns overlays in over a graded frame. So the bridge is a
 * translation, not a new pipeline: Road Trip writes the badge into the
 * project as the scene it always was, and ONE export carries the grade, the
 * telemetry and the hook.
 *
 * Two rules make it safe to run again and again:
 *   - everything injected is named `roadtrip:…` and belongs to the
 *     `roadtrip-hook` scene, so a second send REPLACES the first rather than
 *     stacking a second badge on the clip;
 *   - nothing else in the project is touched. A grade, a trim, the telemetry
 *     elements and the author's own intro survive a send untouched.
 *
 * What does NOT cross over: the shades. Road Trip paints gradients; a Studio
 * scene has a flat scrim, and quietly turning a bottom gradient into a full
 * veil would be a different picture. The strongest shade's colour and strength
 * become the scene's scrim, which is the honest lossy mapping — and it is
 * said out loud in the panel rather than left to be discovered.
 *
 * Pure and DOM-free.
 */

import type { OverlayElement } from '../overlay/overlay-types';
import { OUTRO_SECONDS_DEFAULT, type OutroCard } from '../overlay/outro-card';
import { type Scene, type SceneScrim } from '../overlay/scenes';
import type { ProjectDoc } from '../projects/project-types';
import { isDefaultDevelop, withoutBase, type DevelopSettings } from '../develop/develop';
import { fileBaseName } from '../library/assets';
import { ctaLayout, type CtaSlide } from './cta-slide';
import type { Shade } from './shades';

/** The scene a sent hook lives in. One per project: a clip has one hook. */
export const HOOK_SCENE_ID = 'roadtrip-hook';

/** Every element the bridge owns carries this prefix, and only those. */
export const HOOK_ELEMENT_PREFIX = 'roadtrip:';

export function isHookElement(el: OverlayElement): boolean {
  return el.id.startsWith(HOOK_ELEMENT_PREFIX);
}

/**
 * The flat scrim that stands in for a stack of gradients: the strongest
 * shade's colour at its own strength, or null when there is nothing to stand
 * in for. Lossy on purpose and by exactly one dimension — where the darkening
 * falls — which is the part a scene cannot express.
 */
export function scrimFromShades(shades: readonly Shade[], fade = 0.4): SceneScrim | null {
  let strongest: Shade | null = null;
  for (const shade of shades) {
    if (shade.strength <= 0) continue;
    if (!strongest || shade.strength > strongest.strength) strongest = shade;
  }
  if (!strongest) return null;
  return {
    color: strongest.color,
    // A veil over the whole frame reads far heavier than the same number in a
    // gradient that clears half of it, so it is held back deliberately.
    opacity: Math.min(1, Math.max(0, strongest.strength * 0.6)),
    fade,
  };
}

export interface HookInjection {
  scene: Scene;
  elements: OverlayElement[];
}

/**
 * The badge, translated into a scene and the elements inside it.
 *
 * Element windows become offsets WITHIN the scene (that is what a scene
 * means), so a staggered entrance keeps its stagger and an exit lands on the
 * hook's own duration rather than on the clip's end.
 */
export function hookInjection(
  badgeElements: readonly OverlayElement[],
  durationSeconds: number,
  shades: readonly Shade[] = [],
  name = 'Trip hook',
): HookInjection {
  const end = Number.isFinite(durationSeconds) && durationSeconds > 0 ? durationSeconds : 4;
  return {
    scene: {
      id: HOOK_SCENE_ID,
      name,
      start: 0,
      end,
      scrim: scrimFromShades(shades),
      // The telemetry HUD holds back while the hook runs, then fades in — the
      // studio's own behaviour for an introduction, and the reason the badge
      // and a full HUD do not fight for the same frame.
      solo: true,
      hudFade: 0.5,
    },
    elements: badgeElements.map((el) => ({
      ...el,
      id: `${HOOK_ELEMENT_PREFIX}${el.id}`,
      sceneId: HOOK_SCENE_ID,
    })),
  };
}

/**
 * A project with this hook in it, replacing any hook sent before. Everything
 * the author did in the Studio — the grade, the trim, their own elements and
 * scenes — is left exactly as it was.
 */
export function withHook(doc: ProjectDoc, injection: HookInjection): ProjectDoc {
  const kept = doc.elements.filter((el) => !isHookElement(el));
  const scenes = doc.scenes.filter((s) => s.id !== HOOK_SCENE_ID);
  return {
    ...doc,
    updatedAt: Date.now(),
    // The hook is drawn LAST so it sits over the telemetry, which is what an
    // introduction is for.
    elements: [...kept, ...injection.elements],
    scenes: [...scenes, injection.scene],
  };
}

/** The project with the hook taken back out — and the develop it sent — and nothing else changed. */
export function withoutHook(doc: ProjectDoc): ProjectDoc {
  const sent = Object.entries(doc.media?.develops ?? {}).filter(([, d]) => d.via === 'roadtrip');
  if (!doc.elements.some(isHookElement) && !doc.scenes.some((s) => s.id === HOOK_SCENE_ID) && sent.length === 0) {
    return doc;
  }
  const develops = { ...(doc.media?.develops ?? {}) };
  for (const [key] of sent) delete develops[key];
  return {
    ...doc,
    updatedAt: Date.now(),
    elements: doc.elements.filter((el) => !isHookElement(el)),
    scenes: doc.scenes.filter((s) => s.id !== HOOK_SCENE_ID),
    ...(doc.media ? { media: { ...doc.media, develops } } : {}),
  };
}

// --- the hook picture's DEVELOP --------------------------------------------
//
// The third thing the bridge used to leave behind (`docs/photo-develop.md`
// §7.7, P8 — his "go", 2026-09-24): a reel exported from the linked project
// lacked the correction the badge was composed over. The hook's develop now
// crosses into the project's per-media develops, under the media's key (its
// lowercased base name, the Studio's asset id), guarded by its hash like any
// develop there. Same discipline as the outro: what the bridge writes is
// MARKED (`via: 'roadtrip'`) so a resend replaces only its own, an entry the
// author set in the Studio is never overwritten (it is held and said), and
// an unlink takes only the marked one back out.

/** The hook picture as the bridge sends it: which media, and its correction. */
export interface HookDevelop {
  /** The picture's file name — the project keys a media by its base name. */
  name: string;
  hash?: string | null;
  /** Null or as shot: the project keeps nothing of Trips' for it. */
  settings: DevelopSettings | null;
}

/** The key a project gives a media: its asset id, the lowercased base name. */
export function projectMediaKey(name: string): string {
  return fileBaseName(name).toLowerCase();
}

/**
 * The project with the hook picture's develop written — `held` when the
 * author's own develop for that media is there and was left alone. A
 * develop's RAW base (`base`, `rawGain`, `rawWb`) never crosses: it is a
 * choice about the file Develop opened, and the Studio renders its own.
 */
export function withHookDevelop(
  doc: ProjectDoc,
  hook: HookDevelop | null,
): { doc: ProjectDoc; held: boolean } {
  if (!hook) return { doc, held: false };
  const key = projectMediaKey(hook.name);
  const develops = doc.media?.develops ?? {};
  const there = develops[key];
  if (there && there.via !== 'roadtrip') return { doc, held: true };
  const settings = hook.settings ? withoutBase(hook.settings) : null;
  const next = { ...develops };
  if (settings && !isDefaultDevelop(settings)) {
    next[key] = { settings, ...(hook.hash ? { hash: hook.hash } : {}), via: 'roadtrip' };
  } else if (there) {
    delete next[key];
  } else {
    return { doc, held: false };
  }
  return {
    doc: { ...doc, updatedAt: Date.now(), media: { ...doc.media, develops: next } },
    held: false,
  };
}

// --- the call to action, as the project's OUTRO ----------------------------
//
// The closing card crosses the same bridge the hook does, into the Studio's
// own outro slot (`shared/overlay/outro-card.ts`): the card the carousel
// export appends as its last slide becomes, on a reel, the card the video
// export appends after the footage. Same translation discipline — everything
// injected is `roadtrip:`-prefixed so a resend replaces rather than stacks,
// and an outro the author composed THEMSELVES in the Studio is never touched:
// "nothing else in the project is touched" covers their outro too, and the
// panel says so instead of silently overwriting it.

/** True when a project's outro is one this bridge put there. */
export function isRoadtripOutro(outro: OutroCard | null | undefined): boolean {
  return !!outro && outro.elements.some((el) => el.id.startsWith(HOOK_ELEMENT_PREFIX));
}

/**
 * The trip's call to action as an outro card, laid out for the piece's own
 * frame — the same `ctaLayout` the carousel's closing slide uses, so the two
 * cards are the same card. Null when the CTA has nothing to say: a blank
 * end card is worse than none (the deck's own rule).
 */
export function ctaOutro(
  cta: CtaSlide,
  aspect: number,
  seconds: number = OUTRO_SECONDS_DEFAULT,
): OutroCard | null {
  if (!cta.headline.trim() && !cta.body.trim() && !cta.url.trim()) return null;
  const layout = ctaLayout(cta, aspect);
  const qr = layout.qr
    ? {
        url: cta.url.trim(),
        x: layout.qr.x,
        y: layout.qr.y,
        sizeFrac: layout.qr.sizeFrac,
        dark: cta.ink,
        light: cta.background,
      }
    : null;
  return {
    seconds,
    background: cta.background,
    elements: layout.elements.map((el) => ({
      ...el,
      id: `${HOOK_ELEMENT_PREFIX}${el.id}`,
    })),
    qr,
  };
}

/**
 * A project with this call-to-action outro — replacing one this bridge sent
 * before, or filling an empty slot. `card` null means the author unticked the
 * CTA: a previously sent card is taken back out. Returns null instead of a
 * document when the project carries an outro of its OWN, which a send must
 * not overwrite — the caller says so in the panel.
 */
export function withCtaOutro(
  doc: ProjectDoc,
  card: OutroCard | null,
): ProjectDoc | null {
  // `?? null`: a document from before the field existed reads as empty, the
  // same answer the migration gives it.
  const current = doc.outro ?? null;
  const foreign = current !== null && !isRoadtripOutro(current);
  if (foreign) return card === null ? doc : null;
  if (card === null && current === null) return doc;
  return { ...doc, updatedAt: Date.now(), outro: card };
}

/** The project without a bridge-sent outro; the author's own is left alone. */
export function withoutCtaOutro(doc: ProjectDoc): ProjectDoc {
  if (!isRoadtripOutro(doc.outro ?? null)) return doc;
  return { ...doc, updatedAt: Date.now(), outro: null };
}

/** Whether a project is currently carrying a sent hook. */
export function hasHook(doc: ProjectDoc): boolean {
  return doc.scenes.some((s) => s.id === HOOK_SCENE_ID);
}

File: apple/Packages/AtelierKit/Sources/AtelierKit/Roadtrip/HookScene.swift
// Handing a Road Trip hook to the Studio, as an intro SCENE — and its closing
// card as the project's OUTRO, and its picture's develop. Port of
// `src/shared/roadtrip/hook-scene.ts`.
//
// A translation, not a new pipeline: the badge is already overlay elements,
// the Studio already holds a group of elements for the first seconds of a
// clip (a scene), and its export already burns overlays in over a graded
// frame — so ONE export carries the grade, the telemetry and the hook.
//
// Rules kept:
// - Everything injected is named `roadtrip:…` and belongs to the
//   `roadtrip-hook` scene, so a second send REPLACES the first; nothing else
//   in the project is touched (a grade, a trim, the telemetry, the author's
//   own intro and outro survive a send).
// - The shades do NOT cross: a scene has a flat scrim, so the strongest
//   shade's colour and strength stand in, held back, and the panel says so.
// - The hook picture's DEVELOP crosses under the media's key, marked `via:
//   roadtrip` so a resend replaces only its own; an entry the author set in
//   the Studio is never overwritten (it is `held` and said); a RAW base never
//   crosses; an unlink takes only the marked one back out.
// - The call to action becomes the project's OUTRO through the same
//   `ctaLayout` the carousel's last slide uses; an outro the author composed
//   is never overwritten (`withCtaOutro` answers nil for the panel to say).
// - `updatedAt` is stamped from `now` (ms), a parameter, never the clock.

import Foundation

/// The scene a sent hook lives in. One per project: a clip has one hook.
public let hookSceneId = "roadtrip-hook"

/// Every element the bridge owns carries this prefix, and only those.
public let hookElementPrefix = "roadtrip:"

public func isHookElement(_ el: OverlayElement) -> Bool {
    el.id.hasPrefix(hookElementPrefix)
}

/// The flat scrim that stands in for a stack of gradients: the strongest
/// shade's colour at its own strength, held back, or nil with nothing to
/// stand in for. Lossy on purpose, by exactly where the darkening falls.
public func scrimFromShades(_ shades: [Shade], _ fade: Double = 0.4) -> SceneScrim? {
    var strongest: Shade? = nil
    for shade in shades where shade.strength > 0 {
        if strongest == nil || shade.strength > strongest!.strength { strongest = shade }
    }
    guard let strongest else { return nil }
    // A veil over the whole frame reads far heavier than the same number in a
    // gradient that clears half of it.
    return SceneScrim(color: strongest.color, opacity: min(1, max(0, strongest.strength * 0.6)), fade: fade)
}

public struct HookInjection: Equatable, Sendable {
    public var scene: OverlayScene
    public var elements: [OverlayElement]

    public init(scene: OverlayScene, elements: [OverlayElement]) {
        self.scene = scene; self.elements = elements
    }
}

/// The badge, translated into a scene and the elements inside it. Element
/// windows are offsets WITHIN the scene, so a staggered entrance keeps its
/// stagger and an exit lands on the hook's own duration.
public func hookInjection(_ badgeElements: [OverlayElement], _ durationSeconds: Double, _ shades: [Shade] = [],
                          _ name: String = "Trip hook") -> HookInjection {
    let end = durationSeconds.isFinite && durationSeconds > 0 ? durationSeconds : 4
    // The telemetry HUD holds back while the hook runs, then fades in.
    let scene = OverlayScene(id: hookSceneId, name: name, start: 0, end: end, scrim: scrimFromShades(shades),
                             solo: true, hudFade: 0.5)
    let elements = badgeElements.map { el -> OverlayElement in
        var out = el
        out.id = hookElementPrefix + el.id
        out.sceneId = hookSceneId
        return out
    }
    return HookInjection(scene: scene, elements: elements)
}

/// A project with this hook in it, replacing any hook sent before. The hook is
/// drawn LAST, over the telemetry.
public func withHook(_ doc: ProjectDoc, _ injection: HookInjection, now: Double = nowMillis()) -> ProjectDoc {
    var out = doc
    out.updatedAt = now
    out.elements = doc.elements.filter { !isHookElement($0) } + injection.elements
    out.scenes = doc.scenes.filter { $0.id != hookSceneId } + [injection.scene]
    return out
}

/// The project with the hook taken back out — and the develop it sent — and
/// nothing else changed; the same document when there was nothing to take.
public func withoutHook(_ doc: ProjectDoc, now: Double = nowMillis()) -> ProjectDoc {
    let sent = doc.media.develops.filter { $0.value.via == .roadtrip }.map(\.key)
    if !doc.elements.contains(where: isHookElement) && !doc.scenes.contains(where: { $0.id == hookSceneId })
        && sent.isEmpty {
        return doc
    }
    var out = doc
    out.updatedAt = now
    out.elements = doc.elements.filter { !isHookElement($0) }
    out.scenes = doc.scenes.filter { $0.id != hookSceneId }
    for key in sent { out.media.develops[key] = nil }
    return out
}

// MARK: - the hook picture's develop

/// The hook picture as the bridge sends it: which media, and its correction.
public struct HookDevelop: Equatable, Sendable {
    /// The picture's file name — the project keys a media by its base name.
    public var name: String
    public var hash: String?
    /// Nil or as shot: the project keeps nothing of Trips' for it.
    public var settings: DevelopSettings?

    public init(name: String, hash: String? = nil, settings: DevelopSettings?) {
        self.name = name; self.hash = hash; self.settings = settings
    }
}

/// The key a project gives a media: its asset id, the lowercased base name.
public func projectMediaKey(_ name: String) -> String {
    fileBaseName(name).lowercased()
}

/// What `withHookDevelop` answers: the document, and whether the author's own
/// develop for that media was there and left alone.
public struct HookDevelopWrite: Equatable, Sendable {
    public var doc: ProjectDoc
    public var held: Bool

    public init(doc: ProjectDoc, held: Bool) { self.doc = doc; self.held = held }
}

/// The project with the hook picture's develop written — `held` when the
/// author's own develop for that media is there and was left alone. A RAW
/// base (`base`, `rawGain`, `rawWb`) never crosses.
public func withHookDevelop(_ doc: ProjectDoc, _ hook: HookDevelop?, now: Double = nowMillis()) -> HookDevelopWrite {
    guard let hook else { return HookDevelopWrite(doc: doc, held: false) }
    let key = projectMediaKey(hook.name)
    let there = doc.media.develops[key]
    if let there, there.via != .roadtrip { return HookDevelopWrite(doc: doc, held: true) }
    let settings = hook.settings.map(withoutBase)
    var out = doc
    if let settings, !isDefaultDevelop(settings) {
        let hash = hook.hash.flatMap { $0.isEmpty ? nil : $0 }
        out.media.develops[key] = SavedDevelop(settings: settings, hash: hash, via: .roadtrip)
    } else if there != nil {
        out.media.develops[key] = nil
    } else {
        return HookDevelopWrite(doc: doc, held: false)
    }
    out.updatedAt = now
    return HookDevelopWrite(doc: out, held: false)
}

// MARK: - the call to action, as the project's OUTRO

/// True when a project's outro is one this bridge put there.
public func isRoadtripOutro(_ outro: OutroCard?) -> Bool {
    guard let outro else { return false }
    return outro.elements.contains { $0.id.hasPrefix(hookElementPrefix) }
}

/// The trip's call to action as an outro card, laid out for the piece's own
/// frame by the same `ctaLayout` the carousel's closing slide uses. Nil when
/// the card has nothing to say — a blank end card is worse than none.
public func ctaOutro(_ cta: CtaSlide, _ aspect: Double, _ seconds: Double = outroSecondsDefault) -> OutroCard? {
    let trim = { (s: String) in s.trimmingCharacters(in: .whitespacesAndNewlines) }
    if trim(cta.headline).isEmpty && trim(cta.body).isEmpty && trim(cta.url).isEmpty { return nil }
    let layout = ctaLayout(cta, aspect)
    let qr = layout.qr.map {
        OutroQr(url: trim(cta.url), x: $0.x, y: $0.y, sizeFrac: $0.sizeFrac, dark: cta.ink, light: cta.background)
    }
    let elements = layout.elements.map { el -> OverlayElement in
        var out = el
        out.id = hookElementPrefix + el.id
        return out
    }
    return OutroCard(seconds: seconds, background: cta.background, elements: elements, qr: qr)
}

/// A project with this call-to-action outro — replacing one this bridge sent
/// before, or filling an empty slot; `card` nil takes a sent card back out.
/// Nil instead of a document when the project carries an outro of its OWN,
/// which a send must not overwrite — the caller says so in the panel.
public func withCtaOutro(_ doc: ProjectDoc, _ card: OutroCard?, now: Double = nowMillis()) -> ProjectDoc? {
    let current = doc.outro
    let foreign = current != nil && !isRoadtripOutro(current)
    if foreign { return card == nil ? doc : nil }
    if card == nil && current == nil { return doc }
    var out = doc
    out.updatedAt = now
    out.outro = card
    return out
}

/// The project without a bridge-sent outro; the author's own is left alone.
public func withoutCtaOutro(_ doc: ProjectDoc, now: Double = nowMillis()) -> ProjectDoc {
    if !isRoadtripOutro(doc.outro) { return doc }
    var out = doc
    out.updatedAt = now
    out.outro = nil
    return out
}

/// Whether a project is currently carrying a sent hook.
public func hasHook(_ doc: ProjectDoc) -> Bool {
    doc.scenes.contains { $0.id == hookSceneId }
}
